#VRML_SIM R2023b utf8

PROTO flopsy [
  field SFString controller  "flopsy"
  field SFString name        "flopsy"
  field SFVec3f  translation -0.55 0 0
]
{
  Robot {
    name IS name
    controller IS controller
    translation IS translation
    children [
      Shape {
        appearance PBRAppearance {
          baseColor 0 0 0
          roughness 0.2
          metalness 0
        }
        geometry DEF BODY IndexedFaceSet {
          coord Coordinate {
            point [
              # x axis +right, y axis +forward, z axis +up
              -0.075 0.075 0      # base fl corner
              0.075 0.075 0       # base fr corner
              -0.075 -0.075 0     # base rl corner
              0.075 -0.075 0      # base rr corner
              -0.075 -0.036 0.065 # top left corner
              0.075 -0.036 0.065  # top right corner
            ]
          }
          coordIndex [
            0 1 3 2 -1 # base
            0 2 4 -1   # left side
            1 5 3 -1   # right side
            4 5 1 0 -1 # front slope
            5 4 2 3 -1 # rear slope
          ]
        }
      }
      HingeJoint {
        endPoint Solid {
          name "flipper"
          children [
            DEF FLIPPER_SHAPE Shape {
              appearance PBRAppearance {
                baseColor 0.8 0 0
                roughness 0.2
                metalness 0
              }
              geometry Box {
                size 0.025 0.200 0.005
              }
            }
          ]
          boundingObject USE FLIPPER_SHAPE
          physics Physics {}
          translation 0 0 0.05
          rotation 1 0 0 -0.57
        }
        jointParameters HingeJointParameters {
          anchor 0 0 0.05
          axis 1 0 0
        }
        device RotationalMotor {
          name "flipper_motor"
        }
      }
      HingeJoint {
        endPoint Solid {
          name "fl_wheel"
          children [
            DEF WHEEL_SHAPE Shape {
              appearance PBRAppearance {
                baseColor 0.8 0 0
                roughness 0.2
                metalness 0
              }
              geometry Cylinder {
                radius 0.0155
                height 0.0162
              }
            }
          ]
          boundingObject USE WHEEL_SHAPE
          physics DEF WHEEL_PH Physics {}
          translation -0.076 0.05 0.005
          rotation 0 1 0 1.5708
          contactMaterial "freeWheel"
        }
        jointParameters HingeJointParameters {
          anchor -0.075 0.05 0.005
          axis -1 0 0
        }
      }
      HingeJoint {
        endPoint Solid {
          name "fr_wheel"
          children [ USE WHEEL_SHAPE ]
          boundingObject USE WHEEL_SHAPE
          physics USE WHEEL_PH
          translation 0.076 0.05 0.005
          rotation 0 1 0 1.5708
          contactMaterial "freeWheel"
        }
        jointParameters HingeJointParameters {
          anchor 0.075 0.05 0.005
          axis -1 0 0
        }
      }
      HingeJoint {
        endPoint Solid {
          name "rl_wheel"
          children [ USE WHEEL_SHAPE ]
          boundingObject USE WHEEL_SHAPE
          physics USE WHEEL_PH
          translation -0.076 -0.05 0.005
          rotation 0 1 0 1.5708
          contactMaterial "drivingWheel"
        }
        jointParameters HingeJointParameters {
          anchor -0.075 -0.05 0.005
          axis -1 0 0
        }
        device RotationalMotor {
          name "l_motor"
        }
      }
      HingeJoint {
        endPoint Solid {
          name "rr_wheel"
          children [ USE WHEEL_SHAPE ]
          boundingObject USE WHEEL_SHAPE
          physics USE WHEEL_PH
          translation 0.076 -0.05 0.005
          rotation 0 1 0 1.5708
          contactMaterial "drivingWheel"
        }
        jointParameters HingeJointParameters {
          anchor 0.075 -0.05 0.005
          axis -1 0 0
        }
        device RotationalMotor {
          name "r_motor"
        }
      }
    ]
    boundingObject USE BODY
    physics Physics {}
  }
}